support/magicleap/Servo2D/code/src/Servo2D.cpp

```diff
@@ -0,0 +1,187 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <Servo2D.h>
+#include <lumin/node/RootNode.h>
+#include <lumin/node/QuadNode.h>
+#include <lumin/ui/Cursor.h>
+#include <ml_logging.h>
+#include <scenesGen.h>
+#include <SceneDescriptor.h>
+#include <EGL/egl.h>
+#include <GLES/gl.h>
+#include <string.h>
+
+// A function which calls the ML logger, suitable for passing into Servo
+typedef void (*MLLogger)(MLLogLevel lvl, char* msg);
+void logger(MLLogLevel lvl, char* msg) {
+  if (MLLoggingLogLevelIsEnabled(lvl)) {
+    MLLoggingLog(lvl, ML_DEFAULT_LOG_TAG, msg);
+  }
+}
+
+// The functions Servo provides for hooking up to the ML.
+// For the moment, this doesn't handle input events.
+extern "C" ServoInstance init_servo(EGLContext, EGLSurface, EGLDisplay, MLLogger, const char*);
+extern "C" void heartbeat_servo(ServoInstance);
+extern "C" void discard_servo(ServoInstance);
+
+// Create a Servo2D instance
+Servo2D::Servo2D() {
+  ML_LOG(Debug, "Servo2D Constructor.");
+}
+
+// Destroy a Servo 2D instance
+Servo2D::~Servo2D() {
+  ML_LOG(Debug, "Servo2D Destructor.");
+  discard_servo(servo_);
+  servo_ = nullptr;
+}
+
+// The prism dimensions
+const glm::vec3 Servo2D::getInitialPrismExtents() const {
+  return glm::vec3(0.5f, 0.5f, 0.5f);
+}
+
+// Create the prism for Servo
+int Servo2D::createInitialPrism() {
+  prism_ = requestNewPrism(getInitialPrismExtents());
+  if (!prism_) {
+    ML_LOG(Error, "Servo2D Error creating default prism.");
+    return 1;
+  }
+  return 0;
+}
+
+// Initialize a Servo instance
+int Servo2D::init() {
+
+  ML_LOG(Debug, "Servo2D Initializing.");
+
+  // Set up the prism
+  createInitialPrism();
+  lumin::ui::Cursor::SetScale(prism_, 0.03f);
+  instanceInitialScenes();
+
+  // Get the panar resource that holds the EGL context
+  lumin::RootNode* root_node = prism_->getRootNode();
+  if (!root_node) {
+    ML_LOG(Error, "Servo2D Failed to get root node");
+    abort();
+    return 1;
+  }
+
+  std::string content_node_id = Servo2D_exportedNodes::content;
+  lumin::QuadNode* content_node = lumin::QuadNode::CastFrom(prism_->findNode(content_node_id, root_node));
+  if (!content_node) {
+    ML_LOG(Error, "Servo2D Failed to get content node");
+    abort();
+    return 1;
+  }
+
+  lumin::ResourceIDType plane_id = prism_->createPlanarEGLResourceId();
+  if (!plane_id) {
+    ML_LOG(Error, "Servo2D Failed to create EGL resource");
+    abort();
+    return 1;
+  }
+
+  plane_ = static_cast<lumin::PlanarResource*>(prism_->getResource(plane_id));
+  if (!plane_) {
+    ML_LOG(Error, "Servo2D Failed to create plane");
+    abort();
+    return 1;
+  }
+
+  content_node->setRenderResource(plane_id);
+
+  // Get the EGL context, surface and display.
+  EGLContext ctx = plane_->getEGLContext();
+  EGLSurface surf = plane_->getEGLSurface();
+  EGLDisplay dpy = eglGetDisplay(EGL_DEFAULT_DISPLAY);
+  eglMakeCurrent(dpy, surf, surf, ctx);
+  glViewport(0, 0, 500, 500);
```

Comment:
Does this need to match the 500s in libmlservo? Can we pass this value in to Rust?